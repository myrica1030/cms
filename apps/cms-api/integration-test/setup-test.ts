/* eslint-disable no-console */
import { spawn } from 'node:child_process'
import path from 'node:path'
import process from 'node:process'
import { URL } from 'node:url'
import { config } from 'dotenv-flow'

config({
  files: [path.resolve(__dirname, '../.env')],
})

if (!process.env.DATABASE_URL) {
  throw new Error('DATABASE_URL is not defined!')
}

const databaseUrl = new URL(process.env.DATABASE_URL)
if (!databaseUrl.pathname.endsWith('_test')) {
  databaseUrl.pathname += '_test'
}

process.env.DATABASE_URL = databaseUrl.toString()
process.env.APP_SECRET = 'secret'

async function execute(cmd: string, stdout: boolean = false) {
  return await new Promise((resolve, reject) => {
    const stream = spawn(cmd, {
      shell: true,
      cwd: path.resolve(__dirname, '..'),
      env: { ...process.env },
    })
    if (stdout) stream.stdout.pipe(process.stdout)
    stream.stderr.pipe(process.stderr)
    stream.on('close', code => {
      if (code === 0) resolve(code)
      // eslint-disable-next-line prefer-promise-reject-errors
      else reject()
    })
  })
}

export default async function setup() {
  console.time('Testing database is ready!')
  console.log(`Preparing reset testing database... (${databaseUrl.toString()})`)
  await execute('pnpm prisma migrate reset --force')
  console.log(`Preparing insert example data...`)
  await execute('pnpm db:seed-example')
  console.timeEnd(`Testing database is ready!`)
  console.log('')
}
