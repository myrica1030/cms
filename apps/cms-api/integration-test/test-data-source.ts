import type { TypeOrmModuleOptions } from '@nestjs/typeorm'
import type { DataSourceOptions } from 'typeorm'
import { DataSource } from 'typeorm'
import { allEntities } from 'src/data-source'

export const testDataSourceOptions: DataSourceOptions = {
  type: 'sqlite',
  database: ':memory:',
  entities: allEntities,
  dropSchema: true,
  synchronize: true,
}

export const testDataSource = new DataSource(testDataSourceOptions)

export const testTypeormOptions: TypeOrmModuleOptions = {
  ...testDataSourceOptions,
}
