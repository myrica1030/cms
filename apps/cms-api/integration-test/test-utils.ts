/* eslint-disable ts/no-unsafe-argument,ts/no-unsafe-member-access */
import type { INestApplication } from '@nestjs/common'
import type { TestingModule } from '@nestjs/testing'
import { Test } from '@nestjs/testing'
import { HttpExceptionFilter } from 'common/filter/http-exception.filter'
import { validationPipe } from 'common/pipe/pipes'
import type { Response } from 'supertest'
import request from 'supertest'
import { AppModule } from 'src/app/app.module'

export async function getToken(app: INestApplication): Promise<string> {
  return await new Promise((resolve, reject) => {
    request(app.getHttpServer()).post('/auth/login')
      .send({
        username: 'admin',
        password: '123456',
      })
      .then(response => {
        resolve(response.body.token)
      }, reject)
  })
}

export function mockDate(date: Date | string | number): () => void {
  const { Date } = globalThis

  class MockDate extends Date {
    constructor() {
      super(date) // add whatever date you'll expect to get
    }
  }

  globalThis.Date = MockDate as DateConstructor

  return () => {
    globalThis.Date = Date
  }
}

export async function initIntegrationTestingModule(app: INestApplication): Promise<INestApplication> {
  const moduleFixture: TestingModule = await Test.createTestingModule({
    imports: [AppModule],
  }).compile()

  app = moduleFixture.createNestApplication()
  app.useGlobalPipes(validationPipe)
  app.useGlobalFilters(new HttpExceptionFilter())

  await app.init()
  return app
}

export function prettyResponse(response: Response) {
  return Object.fromEntries(Object.entries(response.body).map(([k, v]) => {
    if (typeof v === 'string' && v.includes('\n')) return [k, v.split('\n')]
    return [k, v]
  }))
}
