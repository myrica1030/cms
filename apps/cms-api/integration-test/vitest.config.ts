import path from 'node:path'
import swc from 'unplugin-swc'
import { defineConfig } from 'vitest/config'

export default defineConfig({
  resolve: {
    alias: {
      'src': path.resolve(__dirname, '../src'),
      'integration-test': path.resolve(__dirname, '.'),
    },
  },
  plugins: [
    swc.vite() as any,
  ],
  test: {
    name: 'api-integration',
    dir: __dirname,
    include: ['specs/**/*.integration.spec.ts'],
    globals: true,
    reporters: ['basic', 'html', 'json', 'junit'],
    coverage: {
      provider: 'v8',
      include: ['src/**/*.ts'],
      exclude: [
        'src/*.ts',
        'src/**/*.spec.ts',
        'src/**/*.{module,dto,ro,strategy,controller,fixture}.ts',
      ],
    },
  },
})
