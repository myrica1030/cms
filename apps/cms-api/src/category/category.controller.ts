import { Body, Controller, Get, NotFoundException, Param, Post } from '@nestjs/common'
import { ApiCreatedResponse, ApiNotFoundResponse, ApiOkResponse, ApiOperation, ApiParam, ApiTags } from '@nestjs/swagger'
import { CategoryService } from 'src/category/category.service'
import { CreateCategoryDto } from 'src/category/dto/create-category.dto'
import { CategoryEntity } from 'src/category/entity/category.entity'
import { ApiInvalidFormResponse } from 'src/decorators'
import { UseJwtGuards } from 'src/guards'

@Controller('category')
@ApiTags('Category')
export class CategoryController {
  constructor(
    private readonly service: CategoryService,
  ) {}

  @UseJwtGuards()
  @ApiOperation({ summary: 'Create a category' })
  @ApiCreatedResponse({ type: CategoryEntity })
  @ApiInvalidFormResponse()
  @Post()
  async createCategory(@Body() createCategoryDto: CreateCategoryDto): Promise<CategoryEntity> {
    const category = await this.service.createCategory(createCategoryDto)
    return new CategoryEntity(category)
  }

  @ApiOperation({ summary: 'Retrieve some categories that not have parent category' })
  @ApiOkResponse({ type: CategoryEntity, isArray: true })
  @ApiNotFoundResponse()
  @Get()
  async retrieveRootCategories(): Promise<CategoryEntity[]> {
    const categories = await this.service.retrieveRootCategories()
    return categories.map(category => new CategoryEntity(category))
  }

  @ApiOperation({ summary: 'Retrieve a category' })
  @ApiParam({ name: 'categoryId', example: 1, type: 'number' })
  @ApiOkResponse({ type: CategoryEntity })
  @ApiNotFoundResponse({ description: 'Category not found' })
  @Get(':categoryId')
  async retrieveCategory(@Param('categoryId') categoryId: number): Promise<CategoryEntity> {
    const category = await this.service.findCategory(categoryId)
    if (!category) throw new NotFoundException()
    return new CategoryEntity(category)
  }
}
