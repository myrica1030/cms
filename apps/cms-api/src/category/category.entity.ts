import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger'
import {
  Column,
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
  Tree,
  TreeChildren,
  TreeParent,
  UpdateDateColumn,
} from 'typeorm'
import { ApiPropertyDatetime, ApiPropertyRichText } from 'src/decorators'

@Entity('category')
@Tree('closure-table')
export class CategoryEntity {
  @PrimaryGeneratedColumn()
  @ApiProperty({ example: 1 })
  id: number

  @TreeParent()
  @ApiPropertyOptional({ type: CategoryEntity, description: 'Category parent' })
  parent?: CategoryEntity

  @TreeChildren()
  @ApiProperty({ type: CategoryEntity, isArray: true, description: 'Category children' })
  children: CategoryEntity[]

  @Column({ type: 'text', unique: true })
  @ApiProperty({ example: 'study-notes' })
  key: string

  @Column({ type: 'text' })
  @ApiProperty({ example: 'Study notes' })
  label: string

  @Column({ type: 'text' })
  @ApiPropertyRichText()
  description?: string

  @CreateDateColumn()
  @ApiPropertyDatetime()
  createdAt: string

  @UpdateDateColumn()
  @ApiPropertyDatetime()
  updatedAt: string
}
