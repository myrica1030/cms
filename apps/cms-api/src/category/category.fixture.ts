import type { Category } from '@prisma/client'
import type { CreateCategoryDto } from 'src/category/dto/create-category.dto'

const dto: CreateCategoryDto = {
  key: 'study-notes',
  label: 'Study notes',
  description: '<p>This is personal study notes</p>',
}

const uncategorizedCategoryEntity: Category = {
  id: 1,
  key: 'uncategorized',
  label: 'Uncategorized',
  description: '',
  createdAt: new Date('2024-09-02T20:04:58+0800'),
  updatedAt: new Date('2024-09-02T20:04:58+0800'),
  parentId: null,
}

const entity: Category = {
  id: 2,
  key: 'study-notes',
  label: 'Study notes',
  description: '<p>This is personal study notes</p>',
  createdAt: new Date('2021-04-18T07:51:33.299Z'),
  updatedAt: new Date('2021-04-18T07:51:33.299Z'),
  parentId: null,
}

const entities: Category[] = [
  uncategorizedCategoryEntity,
  entity,
]

export const categoryFixture = {
  dto,
  uncategorizedCategoryEntity,
  entity,
  entities,
}
