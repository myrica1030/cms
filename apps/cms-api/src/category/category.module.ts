import { Module } from '@nestjs/common'
import { TypeOrmModule } from '@nestjs/typeorm'
import { ArticleEntity } from 'src/article/article.entity'
import { CategoryEntity } from 'src/category/category.entity'
import { CategoryController } from './category.controller'
import { CategoryService } from './category.service'

@Module({
  imports: [
    TypeOrmModule.forFeature([ArticleEntity, CategoryEntity]),
  ],
  controllers: [CategoryController],
  providers: [CategoryService],
})
export class CategoryModule {}
