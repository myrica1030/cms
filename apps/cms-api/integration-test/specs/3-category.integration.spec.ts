import type { INestApplication } from '@nestjs/common'
import { getToken, initIntegrationTestingModule } from 'integration-test/test-utils'
import request from 'supertest'
import { categoryFixture } from 'src/category/category.fixture'
import type { CreateCategoryDto } from 'src/category/dto/create-category.dto'
import type { CategoryEntity } from 'src/category/entity/category.entity'
import { anyDateString } from '../../test-utils/expect.util'

describe('category module', () => {
  let app: INestApplication
  let token: string

  beforeAll(async () => {
    app = await initIntegrationTestingModule(app)
    token = await getToken(app)
  })

  const dto: CreateCategoryDto = {
    key: 'study-notes2',
    label: 'Study notes2',
    description: '<p>This is personal study notes2</p>',
  }

  describe('/category (POST)', () => {
    it('should return 201 when post a valid create category form', async () => {
      const response = await request(app.getHttpServer())
        .post('/category')
        .auth(token, { type: 'bearer' })
        .send(dto)

      expect(response.status, JSON.stringify(response.body)).toBe(201)
      expect(response.body).toEqual({
        id: 3,
        ...dto,
        createdAt: expect.stringMatching(anyDateString),
        updatedAt: expect.stringMatching(anyDateString),
      } satisfies CategoryEntity)
    })

    it('should return 422 when post an existed category key', async () => {
      const response = await request(app.getHttpServer())
        .post('/category')
        .auth(token, { type: 'bearer' })
        .send(categoryFixture.dto)

      expect(response.status).toBe(422)
      expect(response.body).toEqual({ key: ['isExist'] })
    })
  })

  describe.skip('/category (GET)', () => {
    it('should return all categories that without parent category', async () => {
      const response = await request(app.getHttpServer())
        .get('/category')

      expect(response.status).toBe(200)
    })
  })

  describe('/category/:id (GET)', () => {
    it('should return category given existed category id', async () => {
      const response = await request(app.getHttpServer())
        .get('/category/1')

      expect(response.status).toBe(200)
      expect(response.body).toMatchSnapshot()
    })

    it('should return 404 when category not exist', async () => {
      const response = await request(app.getHttpServer())
        .get('/category/999')

      expect(response.status).toBe(404)
    })
  })
})
