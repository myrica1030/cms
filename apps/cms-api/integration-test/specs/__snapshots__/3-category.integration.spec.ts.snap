// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`category module > /category/:id (GET) > should return category given existed category id 1`] = `
{
  "createdAt": "2024-09-02T12:04:58.000Z",
  "description": "",
  "id": 1,
  "key": "uncategorized",
  "label": "Uncategorized",
  "updatedAt": "2024-09-02T12:04:58.000Z",
}
`;
